sycl/source/detail/kernel_data.cpp

```diff
@@ -0,0 +1,367 @@
+                    Index + IndexShift);
+
+  // TODO ESIMD currently does not suport offset, memory and access ranges -
+  // accessor::init for ESIMD-mode accessor has a single field, translated
+  // to a single kernel argument set above.
+  if (!isESIMD && !IsKernelCreatedFromSource) {
+    // Dimensionality of the buffer is 1 when dimensionality of the
+    // accessor is 0.
+    const size_t SizeAccField =
+        sizeof(size_t) * (AccImpl->MDims == 0 ? 1 : AccImpl->MDims);
+    ++IndexShift;
+    Args.emplace_back(kernel_param_kind_t::kind_std_layout,
+                      &AccImpl->MAccessRange[0], SizeAccField,
+                      Index + IndexShift);
+    ++IndexShift;
+    Args.emplace_back(kernel_param_kind_t::kind_std_layout,
+                      &AccImpl->MMemoryRange[0], SizeAccField,
+                      Index + IndexShift);
+    ++IndexShift;
+    Args.emplace_back(kernel_param_kind_t::kind_std_layout,
+                      &AccImpl->MOffset[0], SizeAccField, Index + IndexShift);
+  }
+}
+
+static void addArgsForLocalAccessor(detail::LocalAccessorImplHost *LAcc,
+                                    size_t Index, size_t &IndexShift,
+                                    bool IsKernelCreatedFromSource,
+                                    std::vector<detail::ArgDesc> &Args,
+                                    bool IsESIMD) {
+  using detail::kernel_param_kind_t;
+
+  range<3> &LAccSize = LAcc->MSize;
+  const int Dims = LAcc->MDims;
+  int SizeInBytes = LAcc->MElemSize;
+  for (int I = 0; I < Dims; ++I)
+    SizeInBytes *= LAccSize[I];
+
+  // Some backends do not accept zero-sized local memory arguments, so we
+  // make it a minimum allocation of 1 byte.
+  SizeInBytes = std::max(SizeInBytes, 1);
+  Args.emplace_back(kernel_param_kind_t::kind_std_layout, nullptr, SizeInBytes,
+                    Index + IndexShift);
+  // TODO ESIMD currently does not suport MSize field passing yet
+  // accessor::init for ESIMD-mode accessor has a single field, translated
+  // to a single kernel argument set above.
+  if (!IsESIMD && !IsKernelCreatedFromSource) {
+    ++IndexShift;
+    const size_t SizeAccField = (Dims == 0 ? 1 : Dims) * sizeof(LAccSize[0]);
+    Args.emplace_back(kernel_param_kind_t::kind_std_layout, &LAccSize,
+                      SizeAccField, Index + IndexShift);
+    ++IndexShift;
+    Args.emplace_back(kernel_param_kind_t::kind_std_layout, &LAccSize,
+                      SizeAccField, Index + IndexShift);
+    ++IndexShift;
+    Args.emplace_back(kernel_param_kind_t::kind_std_layout, &LAccSize,
+                      SizeAccField, Index + IndexShift);
+  }
+}
+
+void KernelData::processArg(void *Ptr, const detail::kernel_param_kind_t &Kind,
+                            const int Size, const size_t Index,
+                            size_t &IndexShift, bool IsKernelCreatedFromSource
+#ifndef __INTEL_PREVIEW_BREAKING_CHANGES
+                            ,
+                            bool IsESIMD
+#endif
```

Comment:
Can we update the caller to do

```
  KerneDataObject.IsESIMD = IsESIMD;
  KernelDataObject.proceesArg(/* no IsESIMD parameter*/);
```

instead? Or, in other words, do what we've done with the `DeviceKernelInfoTy` and have extra setting for the ABI compat entry points immediately after crossing the ABI boundary.

Reply:
Unfortunately, we cannot because `KernelData` is not a flat data structure. The `IsESIMD` is stored in the `DeviceKernelInfo` and `KernelData` stores a pointer to it. In case of application is compiled with 6.2 headers the application calls `processArg` before the `DeviceKernelInfo` is assigned and we cannot create the device info at runtime because the kernel name is not assigned yet.

Reply:
I disagree. That should be hacked on the `handler` side (where the problem is) and not here.